File: buerguer-pig/HamburgesasEIngredientes-Java-IvanRC/src/main/java/storage/ingrediente/IngredienteStorageServiceJSON.java
package storage.ingrediente;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;
import config.ConfigApp;
import models.Ingrediente;

import java.io.*;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.List;

public class IngredienteStorageServiceJSON implements IngredienteStorageService{

    ConfigApp configApp = ConfigApp.getInstance();

    File file = new File(configApp.getAPP_DATA()+File.separator+"ingrediente.json");

    public IngredienteStorageServiceJSON() throws IOException {}

    Moshi moshi = new Moshi.Builder().build();

    Type type = Types.newParameterizedType(List.class, Ingrediente.class);
    JsonAdapter<List<Ingrediente>> jsonAdapter = moshi.adapter(type);

    @Override
    public void saveAll(List<Ingrediente> entities) throws IOException {
        String json = jsonAdapter.indent("   ").toJson(entities);
        try(BufferedWriter bw = new BufferedWriter(new FileWriter(file.getPath()))){
            bw.write(json);
        }catch (IOException e){
            System.out.println("Error de escritura en ingrediente.json.");
        }
    }

    @Override
    public List<Ingrediente> loadAll() throws IOException {
        List<Ingrediente> lista = new ArrayList<>();
        if(!file.exists()) return lista;
        try{
            String json = Files.readString(Path.of(file.getPath()));
            lista = (List<Ingrediente>) jsonAdapter.fromJson(json);
        }catch (IOException e){
            System.out.println("Error de lectura en ingrediente.json.");
        }
        return lista;
    }
}

File: buerguer-pig/HamburgesasEIngredientes-Kotlin-IvanRC/src/main/kotlin/storageService/Ingrediente/IngredienteStorageServiceJSON.kt
package storageService.Ingrediente

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import config.ConfigApp
import dto.ingrediente.IngredienteDto
import mapper.toDto
import mapper.toIngrediente
import model.Ingrediente
import mu.KotlinLogging
import java.io.File

@ExperimentalStdlibApi
class IngredienteStorageServiceJSON: IngredienteStorageService {

    private val config = ConfigApp
    private val file = File(config.APP_DATA+ File.separator+"ingredientes.json")
    private val logger = KotlinLogging.logger {}

    val moshi = Moshi.Builder()
        .addLast(KotlinJsonAdapterFactory())
        .build()
    val jsonAdapter = moshi.adapter<List<IngredienteDto>>()

    override fun saveAll(entities: List<Ingrediente>) {
        logger.debug { "Se guardan todos los ingredientes en el fichero JSON." }
        var json = jsonAdapter.indent("   ").toJson(entities.toDto().ingredientesDto)
        file.writeText(json)
    }

    override fun loadAll(): List<Ingrediente> {
        logger.debug { "Se cargán todos los ingredientes del fichero JSON." }
        if(!file.exists()){
            return emptyList()
        }
        val json = jsonAdapter.fromJson(file.readText())!!.map { it.toIngrediente() }
        return json
    }
}